ssl-bootstrap.sh

```diff
@@ -0,0 +1,109 @@
+#!/bin/bash
+
+#Script to create and sign broker certificate for SSL operation. In default mode will
+#generate generic default values for DNs and passwords.
+
+#The following environment should be supplied to Kafka to support this default behavior
+export KAFKA_SSL_KEYSTORE_LOCATION=${KAFKA_SSL_KEYSTORE_LOCATION-/etc/ssl/private/server.keystore.jks}
+export KAFKA_SSL_KEYSTORE_PASSWORD=${KAFKA_SSL_KEYSTORE_PASSWORD-password1234}
+export KAFKA_SSL_KEY_PASSWORD=${KAFKA_SSL_KEY_PASSWORD-password1234}
+export KAFKA_SSL_TRUSTSTORE_LOCATION=${KAFKA_SSL_TRUSTSTORE_LOCATION-/etc/ssl/private/server.truststore.jks}
+export KAFKA_SSL_TRUSTSTORE_PASSWORD=${KAFKA_SSL_TRUSTSTORE_PASSWORD-password1234}
+
+#The following environment is optional
+#export KAFKA_SSL_CLIENT_AUTH=none
+#export KAFKA_SSL_KEYSTORE_TYPE=JKS
+#export KAFKA_SSL_TRUSTSTORE_TYPE=JKS
+export KAFKA_SSL_SECURE_RANDOM_IMPLEMENTATION=${KAFKA_SSL_SECURE_RANDOM_IMPLEMENTATION-SHA1PRNG}
+
+#The following environment is not used by Kafka but controls the creation of bootstrap certificates
+# CA_DN for the DN used to generate the Root CA (in a format supported by openssl)
+# CA_PASSWORD to protect the CA private key
+# CERT_DN for the DN used to generate the server private key (in a format supported by keytool)
+export CA_KEY=${CA_KEY-/etc/ssl/private/ca_key} #The CA Private key used for signing
+export CA_PASSWORD=${CA_PASSWORD-password1234} #The CA Private key password
+export CA_P12_FILE=${CA_P12_FILE-} #Use an outside CA key pair as a PKCS12 file
+export CA_P12_PASSWORD=${CA_PA12_PASSWORD-password1234} #the password for the P12 file
+
+cd /etc/ssl/private
+
+#Check if the Keystore and/or Truststore files are available as Secrets; if they are, then
+#this script does not need to run.
+if [[ -f /run/secrets/server.keystore.jks ]]; then
+   export KAFKA_SSL_KEYSTORE_LOCATION=/run/secrets/server.keystore.jks
+fi
+
+if [[ -f /run/secrets/server.truststore.jks ]]; then
+   export KAFKA_SSL_TRUSTSTORE_LOCATION=/run/secrets/server.truststore.jks
+fi
+
+#If the Keystore file already exists (i.e., has been volume-mounted or loaded as a secret) don't run this script any more
+if [[ ! -f ${KAFKA_SSL_KEYSTORE_LOCATION} ]]; then
+
+if [[ -z "$CERT_DN" ]]; then
+export CERT_DN="cn=$(hostname), ou=GenericOU, o=GenericO, c=US"
```

Comment:
$(hostname) will return the internal docker hostname / docker version-dependent hostname - , i.e. with
```
$ openssl s_client -showcerts -connect localhost:9094 </dev/null

... snip ...
Server certificate
subject=/C=US/O=GenericO/OU=GenericOU/CN=b0c33c7326c1
issuer=/C=US/ST=California/L=Cupertino/CN=b0c33c7326c1
```
What is the strategy for accessing broker from outside the docker network?

Reply:
If relying on autogenerated certificates in this case, the strategy would be to configure the client to ignore certificate issues as in the typical case with self-signed certificates.